examples/java/src/test/java/org/apache/beam/examples/complete/game/GameStatsIT.java

```diff
@@ -0,0 +1,248 @@
+import java.io.IOException;
+import java.util.concurrent.ThreadLocalRandom;
+import org.apache.beam.examples.complete.game.utils.GameConstants;
+import org.apache.beam.runners.direct.DirectOptions;
+import org.apache.beam.sdk.extensions.gcp.options.GcpOptions;
+import org.apache.beam.sdk.io.TextIO;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubClient;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubClient.SubscriptionPath;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubClient.TopicPath;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubIO;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubOptions;
+import org.apache.beam.sdk.io.gcp.testing.BigqueryClient;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.testing.TestPipelineOptions;
+import org.apache.beam.sdk.util.FluentBackoff;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link GameStats}. */
+@RunWith(JUnit4.class)
+public class GameStatsIT {
+  private static final DateTimeFormatter DATETIME_FORMAT =
+      DateTimeFormat.forPattern("YYYY-MM-dd-HH-mm-ss-SSS");
+  private static final String EVENTS_TOPIC_NAME = "events";
+  public static final String GAME_STATS_TEAM_TABLE = "game_stats_team";
+  private static final Integer DEFAULT_ACK_DEADLINE_SECONDS = 60;
+  public static final String SELECT_COUNT_AS_TOTAL_QUERY =
+      "SELECT total_score FROM `%s.%s.%s` where team like(\"AmaranthKoala\")";
+  private GameStatsOptions options =
+      TestPipeline.testingPipelineOptions().as(GameStatsIT.GameStatsOptions.class);
+  @Rule public final transient TestPipeline testPipeline = TestPipeline.fromOptions(options);
+  private static String pubsubEndpoint;
+  private @Nullable ManagedChannel channel = null;
+  private @Nullable TransportChannelProvider channelProvider = null;
+  private @Nullable TopicAdminClient topicAdmin = null;
+  private @Nullable SubscriptionAdminClient subscriptionAdmin = null;
+  private @Nullable TopicPath eventsTopicPath = null;
+  private @Nullable SubscriptionPath subscriptionPath = null;
+  private String projectId;
+  private static final String TOPIC_PREFIX = "gamestats-";
+  private BigqueryClient bqClient;
+  private final String OUTPUT_DATASET = "game_stats_e2e";
+
+  public interface GameStatsOptions extends TestPipelineOptions, GameStats.Options {};
+
+  @Before
+  public void setupTestEnvironment() throws Exception {
+    PipelineOptionsFactory.register(TestPipelineOptions.class);
+    projectId = TestPipeline.testingPipelineOptions().as(GcpOptions.class).getProject();
+
+    setupBigQuery();
```

Comment:
Can you look into using ExamplesUtils for the setup/breakdown?
https://github.com/apache/beam/blob/master/examples/java/src/main/java/org/apache/beam/examples/common/ExampleUtils.java

Reply:
About the setup/breakdown of the BigQuery dataset, which part from ExampleUtils should I take a look? The test requires only the dataset and WriteWindowedToBigQuery provides the rest of the configs needed.

Reply:
I've tried setting up PubSub and BQ using ExampleUtils but neither the topic/subscription nor the dataset get created quickly enough for the reading/writing to take place.
The way they're being setup/teared down right now work as expected. Is it ok if I stick with that approach?

Reply:
It at least doesn't make sense to have the same code in multiple IT files. Do you know what the difference is in the setup for ExamplesUtils? How much longer does it take to be created? Can we use a check with a backoff if the dataset and topics exist before running the pipeline?